### PR TITLE
Ropes authentication

### LINKED ISSUE #11: Authentication
#### As an application hoster I want to add user authentication to my application.

**Story**:

We have done a great job providing immediate value to our customers. At this
point they can both add and delete images, in addition to a few other features
that are either completed or in the works. Along with a growth in our web
service's popularity have come an increase in number of reports of missing
images. Upon investigation, it turns out that the majority of delete requests
are issued by only a handful of IP addresses. While we could blacklist those IP
addresses from the delete operation, it has always been our plan to only permit
the user who added an image to delete that image. To begin that effort we must
first add the concept of a user.

**Acceptance criteria**:
- [x] Users can sign up.
- [x] Users can log in after they've signed up.
- [x] Users can log out after logging in.
- [x] Users can choose to be remembered after logging in.<sup>†</sup>
- [x] Multiple browsers can concurrently hold permanent cookies for the same
  user.
- [x] Logged in users should not be able to register, nor log in again.

<sup>†</sup>: "Remembering" means that users will remain logged into the site
even after completely closing their web browser.

**Note**: Do not use Devise for this task.

**Discussion topics**:
- [x] If we wanted to add more advanced features such as email activation and
  "forgot my password", should we continue to "roll our own" authentication?

**Reference material**:
- https://www.railstutorial.org/book/log_in_log_out

**Further reading**:
- https://github.com/plataformatec/devise

**Dependencies**:
- Delete Image

### PR DESCRIPTION
**Pull request by kkajla12**
Closes #11

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+  private
+
+  def login_params
+    params.require(:login_form).permit(:email, :password, :remember_me)
+  end
+
+  def handle_valid_login(user:)
+    log_in user
+    params[:login_form][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: It'd be nice to see the remember_me part split out into its own commit since it's a decent amount of extra logic.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,12 +1,17 @@
 class UsersController < ApplicationController
+  include Authenticable
+
+  before_action :redirect_if_already_logged_in, only: [:new, :create]
```
Comment (bboe) [inline app/controllers/users_controller.rb]: ⭐️

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
+class SessionsController < ApplicationController
+  include Authenticable
+
+  before_action :redirect_if_already_logged_in, only: [:new, :create]
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: ⭐️

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+  def new
+    @login_form = LoginForm.new
+  end
+
+  def create
+    @login_form = LoginForm.new(login_params)
+    user = @login_form.valid? ? User.find_by(email: params[:login_form][:email].downcase) : nil
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: Instead of using `params[...][...]` what do you think about using the attributes from the form object?

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    if user && user.authenticate(params[:login_form][:password])
+      handle_valid_login(user: user)
+    else
+      handle_invalid_login
+    end
+  end
+
+  private
+
+  def login_params
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: Is this needed given that you have a form object?

File: test/controllers/sessions_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        remember_me: '0'
+      }
+    }
+    assert_response :unprocessable_entity
+
+    assert_equal 'The email and password combination you entered was invalid', flash[:error]
+    assert_select '.login_form_email span', "can't be blank"
+  end
+
+  test 'should show error when attempting login with blank password' do
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: Does this add much value over `'should show error when attempting login with blank email'`?

File: test/controllers/sessions_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        password: 'invalid',
+        remember_me: '0'
+      }
+    }
+    assert_response :unprocessable_entity
+
+    assert_equal 'The email and password combination you entered was invalid', flash[:error]
+  end
+
+  test 'should show only the flash message when attempting login as non-existent user' do
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: Is the `only` condition actually tested for?

File: test/controllers/sessions_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  end
+
+  test 'login without remembering' do
+    log_in_as(user: @user, remember_me: '0')
+    assert_equal @user.id, session[:user_id]
+    assert_nil cookies['user_id']
+    assert_nil cookies['remember_token']
+  end
+
+  test 'redirected to root if authenticated user visits login path' do
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: ⭐️

File: test/controllers/sessions_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    log_in_as(user: @user)
+
+    get login_path
+    assert_redirected_to root_url
+    assert_equal 'You are already logged in', flash[:error]
+  end
+
+  test 'redirected to root if authenticated user attempts to login' do
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: ⭐️

File: test/controllers/sessions_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+  private
+
+  def log_in_as(user:, remember_me: 0)
+    post login_path, params: {
+      login_form: {
+        email: user.email,
+        password: user.password,
+        remember_me: remember_me
+      }
+    }
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: Should there be an assertion to ensure that this works properly?

File: test/controllers/users_controller_test.rb [hunk 11]
```diff
@@ -144,4 +145,48 @@ class UsersControllerTest < ActionDispatch::IntegrationTest
 
     assert_select 'span', "doesn't match Password"
   end
+
+  test 'redirected to root if authenticated user visits signup path' do
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: ⭐️

File: test/controllers/users_controller_test.rb [hunk 12]
```diff
@@ -144,4 +145,48 @@ class UsersControllerTest < ActionDispatch::IntegrationTest
...
+    create_and_login_user
+
+    get signup_path
+    assert_redirected_to root_url
+    assert_equal 'You are already logged in', flash[:error]
+  end
+
+  test 'redirected to root if authenticated user attempts to signup' do
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: ⭐️

File: test/controllers/users_controller_test.rb [hunk 13]
```diff
@@ -144,4 +145,48 @@ class UsersControllerTest < ActionDispatch::IntegrationTest
...
+      password_confirmation: @valid_password
+    })
+    assert_redirected_to root_url
+    assert_equal 'You are already logged in', flash[:error]
+  end
+
+  private
+
+  def create_and_login_user
+    post users_path(user: {
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Can you create a user in a more efficient manner?

This demonstrates the downside to permitting multiple requests in a controller.

File: test/flow_test_helper.rb [hunk 14]
```diff
@@ -16,6 +16,10 @@ class FlowTestCase < ActiveSupport::TestCase
   include Rails.application.routes.url_helpers
 
   fixtures :all
+
+  setup do
+    Capybara.reset_sessions!
```
Comment (bboe) [inline test/flow_test_helper.rb]: While I know why this is needed, there is no context for why to explain to someone who doesn't know. Adding such context would be a good thing to add in its own commit to explain why it's useful, or at least a comment to explain it's purpose (commit messages are better though).

File: test/flows/sessions_crud_test.rb [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+class SessionsCrudTest < FlowTestCase
+  setup do
+    @user = User.create!(name: 'John Doe',
+                         email: 'johndoe@email.com',
+                         password: 'password',
+                         password_confirmation: 'password')
+  end
+
+  test 'user login without remember me' do
```
Comment (bboe) [inline test/flows/sessions_crud_test.rb]: Are these flows? Or to ask another way, what benefit does this provide over your controller tests?

File: test/controllers/sessions_controller_test.rb [hunk 16]
```diff
@@ -72,14 +72,42 @@ class SessionsControllerTest < ActionDispatch::IntegrationTest
...
+
+    delete logout_path
+    assert_redirected_to root_url
+    follow_redirect!
+
+    assert_select 'a[href=?]', login_path
+    assert_select 'a[href=?]', logout_path, count: 0
+  end
+
+  test 'login with valid information followed by simulated multi-browser logout' do
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: What's the purpose of this test?

File: test/helpers/application_helper_test.rb [hunk 17]
```diff
@@ -2,26 +2,27 @@
 
 class ApplicationHelperTest < ActionView::TestCase
   test 'should return active navbar link if current_page is home' do
-    result = '<li class="nav-item active"><a class="nav-link" href="/">Home</a></li>'
+    result = '<li class="nav-item active"><a class="nav-link" data-method="get" href="/">Home</a></li>'
```
Comment (kkajla12) [inline test/helpers/application_helper_test.rb]: Figured it out. This is applied by the `link_to` helper, but it is manually specified by me in my application helper for generating navbar links. I needed to start passing it as an argument to my application helper so that I could specify that clicking the 'logout' button should send a `delete` request to the sessions controller. However, since login/logout haven't actually been implemented at this point, this change doesn't belong here. It belongs after those actions are implemented.

File: app/controllers/sessions_controller.rb (hunk 5 above)
Comment (kkajla12) [inline app/controllers/sessions_controller.rb]: It's not, but I just think using it and being able to pass it's result to the form object makes the code a lot cleaner.

File: test/controllers/sessions_controller_test.rb (hunk 16 above)
Comment (kkajla12) [inline test/controllers/sessions_controller_test.rb]: This test is intended to simulate one browser tab in a situation where the user also has the site open in another tab, is authenticated in both tabs, and logs out in the other tab.

Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: This unfortunately doesn't address the multi-browser part since it's still a single browser. That feature is not something you can easily test automatically. I suggest testing in manually.

File: Gemfile [hunk 18]
```diff
@@ -74,3 +74,6 @@ gem 'mocha'
 
 # Email Validation
 gem 'email_validator'
+
+# Password Hashing
+gem 'bcrypt'
```
Comment (kkajla12) [inline Gemfile]: Didn't see it. Deleted this line and uncommented line 22 in 0f3fb3371555dc77de03aff48c9dcacf88e4750a

File: test/models/user_test.rb [hunk 19]
```diff
@@ -0,0 +1,121 @@
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
+  setup do
```
Comment (kkajla12) [inline test/models/user_test.rb]: Added helper methods for resource/form object creation in `TestCase` and refactor all tests to use these methods in #67

File: test/models/user_test.rb [hunk 20]
```diff
@@ -118,4 +118,13 @@ class UserTest < ActiveSupport::TestCase
     assert_predicate user, :invalid?
     assert_equal "doesn't match Password", user.errors.messages[:password_confirmation].first
   end
+
+  test 'authenticated? should return false for a user with nil digest' do
```
Comment (kkajla12) [inline test/models/user_test.rb]: Yes, it does. I will pull this out into the correct commit and handle all history changing comments after I take care of your other comments.

File: test/models/user_test.rb [hunk 21]
```diff
@@ -0,0 +1,121 @@
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
+  setup do
+    @valid_name = 'John Doe'
+    @invalid_name = 'a' * 51
+    @valid_email = 'johndoe@email.com'
+    @invalid_email = 'johndoe'
+    @long_email = 'a' * 247 + 'email.com'
```
Comment (kkajla12) [inline test/models/user_test.rb]: See https://github.com/AppFolioOnboarding/image-sharer-kkajla12/pull/57#discussion_r85990482

File: test/models/user_test.rb [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+    @valid_password = 'password'
+    @invalid_password = 'pass'
```
Comment (kkajla12) [inline test/models/user_test.rb]: Updated this test in 1249ef8aa42973bc5d0b7e5caa5bf33a139befeb

File: test/models/user_test.rb [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+  end
+
+  test 'name must be present' do
+    user = User.new(email: @valid_email, password: @valid_password, password_confirmation: @valid_password)
+
+    assert_predicate user, :invalid?
+    assert_equal "can't be blank", user.errors.messages[:name].first
+  end
+
+  test 'name length must be less than minimum' do
```
Comment (kkajla12) [inline test/models/user_test.rb]: Updated this test to check both cases in 1249ef8aa42973bc5d0b7e5caa5bf33a139befeb

File: test/models/user_test.rb [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+    assert_equal "can't be blank", user.errors.messages[:email].first
+  end
+
+  test 'email must be unique' do
+    User.create!(name: @valid_name,
+                 email: @valid_email,
+                 password: @valid_password,
+                 password_confirmation: @valid_password)
+
+    user = User.new(name: @valid_name,
```
Comment (kkajla12) [inline test/models/user_test.rb]: Added this test in 1249ef8aa42973bc5d0b7e5caa5bf33a139befeb

File: test/models/user_test.rb [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+                    email: @valid_email,
+                    password: @valid_password,
+                    password_confirmation: @valid_password)
+
+    assert_predicate user, :invalid?
+    assert_equal 'has already been taken', user.errors.messages[:email].first
+  end
+
+  test 'email length is invalid' do
```
Comment (kkajla12) [inline test/models/user_test.rb]: Updated this test to check both cases in 1249ef8aa42973bc5d0b7e5caa5bf33a139befeb

File: test/controllers/users_controller_test.rb [hunk 26]
```diff
@@ -0,0 +1,147 @@
+require 'test_helper'
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  setup do
+    @valid_name = 'John Doe'
```
Comment (kkajla12) [inline test/controllers/users_controller_test.rb]: See https://github.com/AppFolioOnboarding/image-sharer-kkajla12/pull/57#discussion_r85990482. This one is resolved in https://github.com/AppFolioOnboarding/image-sharer-kkajla12/pull/57#discussion_r85990482

File: test/controllers/users_controller_test.rb (hunk 13 above)
Comment (kkajla12) [inline test/controllers/users_controller_test.rb]: Resolved in 86499f27a5b0c1c349d14de8b0faccedb4a46084.

File: test/controllers/users_controller_test.rb [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+        password: @valid_password,
+        password_confirmation: @valid_password
+      })
+    end
+    assert_response :unprocessable_entity
+
+    assert_select 'span', "can't be blank"
+  end
+
+  test 'should not create user with invalid name' do
```
Comment (kkajla12) [inline test/controllers/users_controller_test.rb]: Removed redundant controller tests in 86499f27a5b0c1c349d14de8b0faccedb4a46084

File: test/controllers/sessions_controller_test.rb (hunk 16 above)
Comment (kkajla12) [inline test/controllers/sessions_controller_test.rb]: Removed this test in 5ea0aa6f7bcf92b33bb73db9121b7cdc27dbedfa

File: test/controllers/sessions_controller_test.rb (hunk 10 above)
Comment (kkajla12) [inline test/controllers/sessions_controller_test.rb]: Yup, need to check the response status. Resolved in 86499f27a5b0c1c349d14de8b0faccedb4a46084

File: test/flows/sessions_crud_test.rb (hunk 15 above)
Comment (kkajla12) [inline test/flows/sessions_crud_test.rb]: I actually argued to not add these flow tests because the controller tests and authenticable tests were checking all of the key functionality. However, Yunlei (who I was pairing with at the time) argued that flow tests would be good in this scenario because they provide and end-to-end check of the authentication.

File: app/controllers/sessions_controller.rb (hunk 4 above)
Comment (kkajla12) [inline app/controllers/sessions_controller.rb]: I like it. Resolved in 86499f27a5b0c1c349d14de8b0faccedb4a46084

**Comment by bboe** [comment]
Go a head and rebase plus squash the appropriate commits before I make the next pass at this one please.

File: test/flow_test_helper.rb (hunk 14 above)
Comment (kkajla12) [inline test/flow_test_helper.rb]: Separated this Capybara configuration into its own with a commit that explains why it is needed. See 1db0f763afd9b7b841c38eeb53076fe835fb0193

File: db/schema.rb [hunk 28]
```diff
@@ -43,4 +43,13 @@
     t.index ["name"], name: "index_tags_on_name", unique: true
   end
 
+  create_table "users", force: :cascade do |t|
+    t.string   "name"
+    t.string   "email"
```
Comment (bboe) [inline db/schema.rb]: It looks like your migration was updated, but not the schema.rb file. I'm hoping it's updated elsewhere, but those updates should probably be here.

File: test/models/user_test.rb [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+    user = new_user(email: email.upcase)
+
+    assert_predicate user, :valid?
+    assert_equal email.upcase, user.email
+
+    user.save!
+    assert_equal email, user.email
+  end
+
+  test 'uppercase duplicate email is still considered duplicate' do
```
Comment (bboe) [inline test/models/user_test.rb]: ⭐️

File: test/test_helper.rb [hunk 30]
```diff
@@ -19,5 +19,19 @@ def create_image!(url: 'https://test.com', tag_list: '0, 1, 2', created_at: Time
     def new_share_form(recipient: 'test@test.com', message: 'Hello, World!')
       ShareImageForm.new(recipient: recipient, message: message)
     end
+
+    def new_user(name: 'John Doe',
```
Comment (bboe) [inline test/test_helper.rb]: This is fine for now, but be cautious in the future as this helper may grow out of control.

File: test/test_helper.rb [hunk 31]
```diff
@@ -20,6 +20,18 @@ def new_share_form(recipient: 'test@test.com', message: 'Hello, World!')
       ShareImageForm.new(recipient: recipient, message: message)
     end
 
+    def user_params(name: 'John Doe',
```
Comment (bboe) [inline test/test_helper.rb]: ⭐️  nice!

File: lib/authenticable.rb [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+  def log_in(user)
+    session[:user_id] = user.id
+  end
+
+  def current_user
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user && user.authenticated?(cookies[:remember_token])
```
Comment (bboe) [inline lib/authenticable.rb]: At this point in the commit history I don't think `remember_token` should exist, but it's hard to tell because of the inconsistent order.

File: lib/authenticable.rb [hunk 33]
```diff
@@ -6,7 +6,14 @@ def log_in(user)
   end
 
   def current_user
-    @current_user ||= User.find_by(id: session[:user_id])
+    if (user_id = session[:user_id])
```
Comment (bboe) [inline lib/authenticable.rb]: Is this as efficient as it used to be?

**Comment by bboe** [comment]
Fixup that one change then merge! Great work!
